### PR TITLE
Implement arbitrary grouping

### PR DESCRIPTION
**Pull request by beorn7**
This fixes https://github.com/prometheus/pushgateway/issues/18 and https://github.com/prometheus/pushgateway/issues/32 .

@brian-brazil @juliusv @fabxc

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -108,42 +129,74 @@ could attach the time of pushing as a timestamp.)
...
+delimited binary protocol buffers or in the simple flat text format
+(both in version 0.0.4, see the
+[data exposition format specification](https://docs.google.com/document/d/1ZjyKiKxZV83VI9ZKAXRGKaUKK2BIWCT7oiGBKDBpjEY/edit?usp=sharing)).
+Discrimination between the two variants is done via the content-type
+header. (In case of an unknown content-type, the text format is tried
+as a fall-back.)
+
+The response code upon success is always 202 (even if the same
+grouping key has never been used before, i.e. there is no feedback to
+the client if the push has replaced a group of metric or created a new
```
Comment (brian-brazil) [inline README.md]: "replaced an existing group"

File: README.md [hunk 2]
```diff
@@ -51,15 +50,15 @@ packet, will result in a protocol error.*
 
 Examples:
 
-* Push a single sample:
+* Push a single sample into the group identified by `{job="some_job}`:
```
Comment (brian-brazil) [inline README.md]: We should mention that `job` is required.

**Comment by brian-brazil** [comment]
Yay!

File: README.md [hunk 3]
```diff
@@ -108,42 +129,74 @@ could attach the time of pushing as a timestamp.)
...
+
+`PUT` is used to push a group of metrics. All metrics with the
+grouping key specified in the URL are replaced by the metrics pushed
+with `PUT`.
+
+The body of the request contains the metrics to push either as
+delimited binary protocol buffers or in the simple flat text format
+(both in version 0.0.4, see the
+[data exposition format specification](https://docs.google.com/document/d/1ZjyKiKxZV83VI9ZKAXRGKaUKK2BIWCT7oiGBKDBpjEY/edit?usp=sharing)).
+Discrimination between the two variants is done via the content-type
```
Comment (juliusv) [inline README.md]: `Content-type`.

File: README.md [hunk 4]
```diff
@@ -108,42 +129,74 @@ could attach the time of pushing as a timestamp.)
...
+header. (In case of an unknown content-type, the text format is tried
```
Comment (juliusv) [inline README.md]: Dito

File: README.md (hunk 1 above)
Comment (juliusv) [inline README.md]: "of metric" -> "of metrics"?

Comment (beorn7) [inline README.md]: Both fixed.

File: README.md (hunk 3 above)
Comment (beorn7) [inline README.md]: Actually `Content-Type`.

File: README.md (hunk 4 above)
Comment (beorn7) [inline README.md]: One could argue that should be "content type"... but I reworded it to refer to the header name, too.

**Comment by beorn7** [comment]
Good to go?

**Comment by beorn7** [comment]
I think @brian-brazil  likes it, but @juliusv should give thumbs up on the code.

**Comment by fabxc** [comment]
Code looked good to me given that it works as expected ;)

**Comment by juliusv** [comment]
Dito :+1:

**Pull request merged**